p4-16/spec/P4-16-spec.mdk

```diff
@@ -7200,6 +7200,63 @@ The `@match` annotation, described in Section [#sec-select], is used
 to specify a `match_kind` value other than the default `match_kind` of
 `exact` for a field of a `value_set`.
 
+### Extern function/method annotations { #sec-extern-annotations }
+
+Various annotations may appear on extern function and method declarations
+to describe limitations on the behavior and interactions of those
+functions.  By default extern functions might have any effect on the
+environment of the P4 program and might interact in non-trivial ways.
+Since externs are architecture-specific and their behavior is known to
+the architecture definition, these annotations are not strictly necessary
+(an implementation can have knowledge of how externs interact based
+on their names built into it), but these annotations provide a uniform
+way of describing certain well-defined interactions (or their absence),
+allowing architecture-indepedent analysis of P4 programs.
+
+- `@pure` - Describes a function that depends solely on its `in` argument
+  values and has no effect other than its `out` argument values and return
+  value.  No hidden state is recorded between calls, and its value does
+  not depend on any hidden state that may be changed by other calls.
+  An example is the `hash` function which computes a deterministic
+  hash of its arguments.  A `@pure` function whose results are unused
+  may be safely eliminated with no adverse effects, and mulitple calls
+  with identical arguments may be combined into a single call.  `@pure`
+  functions may also be reoerdered with respect to other computations
+  that are not data dependent.
+
+- `@noSideEffects` - Weaker than `@pure` and describes a function that
+  does not change any hidden state, but may depend on hidden state.
+  Such a function may be dead code eliminated, and may be reordered
+  or combined with other `@noSideEffects` or `@pure` calls, but not
+  with other function calls that may have side effects that affect
+  the function.
+
+- `@localState` - When applied to an extern method, the extern method
+  implementation might read and/or write internal state, but all such
+  accessed state is limited to state associated directly with the extern
+  object instance on which the method is called. When applied to an extern
+  function, the accessible state is restricted to state associated directly
+  with this extern function. Method and function calls so annotated will
```

Comment:
I suspect that for extern functions you imply that they operate on a "singleton" instance of an implicit built-in extern?

Reply:
That is the meaning I was trying to convey by the sentence "When applied to an extern function, the accessible state is restricted to state associated directly with this extern function."

Reply:
Yes, that is essentially what I was trying to say here.  Not clear if introducing the concept of an implicit singleton makes it clearer or less clear though.